tests/python_package_test/test_sklearn.py

```diff
@@ -1168,6 +1274,7 @@ def fit_and_check(eval_set_names, metric_names, assumed_iteration, first_metric_
         "verbose": -1,
         "seed": 123,
         "early_stopping_rounds": 5,
+        "early_stopping": True,
```

Comment:
Yes, this was a necessary change in order to make the test pass.

With the code in the current state, passing `early_stopping_rounds` while leaving `early_stopping` to its default value (`False`) does not enable early stopping. Before it did.

If we want to keep the previous behaviour and we want to enable auto early stopping when `early_stopping_rounds` is set to a value and `early_stopping` is left to its default value we need to set the default value of `early_stopping_rounds` to `None` and to change it in `process_parameters` to `10` when `early_stopping` is set to `True`, which will make the actual default value of `early_stopping_rounds` a bit hidden in the code.  Let me know what you prefer, both options have their own drawbacks.

Reply:
Ok, that's a problem. And exactly why I have been so strict in asking that we only see **new** tests added in this PR.

I think that also answers my questions from https://github.com/microsoft/LightGBM/pull/5808#discussion_r1885538801

If the current state of this branch were merged, with this mix of behaviors:

* `early_stopping` keyword argument defaulting to `False`
* `early_stopping=False` meaning "do not perform **any** early stopping"

it would **turn off early stopping for all existing code using early stopping with `lightgbm.sklearn` estimators**. 

That's a very big user-facing breaking change, and one I do not support. Please, let's keep the scope of this PR limited to what I've written in the title... "add scikit-learn-style API for early stopping".

Specifically, this is the behavior I'm expecting:

* `early_stopping=True`:
  - *adds a new validation set to `eval_set` which is a random sample of `{validation_fraction} * num_rows` rows from the training data*
  - *enables early stopping, with `early_stopping_rounds` defaulting to 10 if not provided by any other mechanism*
  - *does not affect any other relevant early stopping behavior... e.g. if you also passed a dataset to `eval_set` in `fit(...)`, then that dataset is ALSO used for early stopping*
* `early_stopping=False`
  - *everything `lightgbm` does is identical to what it does today*

The docstring for the `early_stopping` keyword argument should also be updated to clearly explain this. For example:

```text
        early_stopping : bool, optional (default=False)
            Whether to enable scikit-learn-style early stopping. If set to ``True`, 
            a new validation set will be created by randomly sampling ``validation_fraction`` rows
            from the training data ``X`` passed to ``fit()``. Training will stop if the validation score
            does not improve for a specific number of rounds (controlled by ``n_iter_no_change``).
            This parameter is here for compatibility with ``scikit-learn``'s ``HistGradientBoosting``
            estimators... it does not affect other ``lightgbm``-specific early stopping mechanisms,
            like passing the ``lgb.early_stopping`` callback and validation sets to the ``eval_set``
            argument of `fit()`.
```

> *If we want to keep the previous behaviour and we want to enable auto early stopping when `early_stopping_rounds` is set to a value and `early_stopping` is left to its default value we need to set the default value of `early_stopping_rounds` to `None`*

I think this statement is also coming from you trying to tie this new `.early_stopping` property of the `scikit-learn` estimators to "whether or not *any* early stopping happens", and I guess you are referring to this:

https://github.com/microsoft/LightGBM/blob/4feee28a53b7b04e66ee09103d2aac260404b9cf/python-package/lightgbm/engine.py#L275-L288

I don't think we should do that. Let's limit the `early_stopping` keyword argument to the narrow purpose of "whether or not to add 1 randomly-sampled validation set and perform early stopping on it".

Then, the only thing that has to be figured out is how many rounds to to train without improvement before stopping. I gave a proposal for doing that here: https://github.com/microsoft/LightGBM/pull/5808#discussion_r1885546892

Reply:
@ClaudioSalvatoreArcidiacono, I'd like to propose that we take a different approach to getting this PR done.

Could I take over this PR and push to it, to get it into a state that I'd feel comfortable merging?

Then:

1. *you* review the state and tell me if anything that's left concerns you
2. after you and I agree on the state, we ask other maintainers to review

This has been taking a really large amount of both your and my time and energy. I think this approach would get us to a resolution faster.

I would only push new commits and merge commits (no force-pushing), so you could see everything I changed and revert any of it.

Can we do that?

Reply:
Hey @jameslamb, the statement

> If the current state of this branch were merged, with this mix of behaviours:
>
> - early_stopping keyword argument defaulting to False
> - early_stopping=False meaning "do not perform any early stopping"
>
> it would turn off early stopping for all existing code using early stopping with lightgbm.sklearn estimators.

Is not correct. If the current state of this branch were merged it would not enable early stopping **automatically**. Meaning that passing an `early_stopping` callback while leaving early_stopping keyword argument defaulting to False will still enable early stopping as it has always had. See for example [this other test](https://github.com/ClaudioSalvatoreArcidiacono/LightGBM/blob/3313-enable-auto-early-stopping/tests/python_package_test/test_sklearn.py#L172) which is left unchanged and it passes.

The only non backward compatible change introduced at this stage is passing early_stopping_rounds is not sufficient anymore to enable early stopping (if the early_stopping parameter is left to False). I am definetly open to change that.

Regarding your last suggestion, I am extremely thankful for the effort you have invested in this PR and I am sorry to hear that it is taking more effort than you have anticipated. I am putting my best intentions into this PR and I definetly do not want cause any harm to the library or waste any of your precious time unnecessarily.

I would like to give this PR one last attempt, addressing all of your comments and making sure everything is spot on. I will tag you once I think the PR is ready for a last review. If you think that after this last attempt the gap to be filled is still to big, I will step aside and I will happily let you take over and review your code. 

Are you fine with it :)?

Reply:
@jameslamb, In your previous comment you mentioned

> - early_stopping=True:
>   - adds a new validation set to eval_set which is a random sample of {validation_fraction} * num_rows rows from the training data
>   - enables early stopping, with early_stopping_rounds defaulting to 10 if not provided by any other mechanism
>   - does not affect any other relevant early stopping behavior... e.g. if you also passed a dataset to eval_set in fit(...),  then that dataset is ALSO used for early stopping.

Do you think that we should create an extra validation set even if a validation set has been provided in the `fit()`?
I think that if a validation set has been provided in the `fit()` and the parameter `early_stopping=True` then we do not need to create a new validation set for early stopping. What do you think?